File: agent-driven-architecture-main/examples/kotlin/spine/src/main/kotlin/adr/spine/pure/StepRecord.kt
// ── spine/pure/step-record — THE unit of commit and THE unit of replay (G9) ─
// The shipped reference committed `append(signedCommands, capturedResults)` — no
// clock. A live boundary that folded at now=1001 re-folded at now=0, because the
// only thing that could have carried the timestamp was never written down.
//
// The commit is the STEP, not the pair. `now` rides the record, or a re-fold
// cannot reproduce what a live boundary wrote.
//
// Every field earns its place:
//   schemaVersion  which shape the rest of the record is in (14.7) — the envelope,
//             so a log written across deployments is self-describing
//   now       the injected clock read — without it, timestamps are lost outright
//   sig       the stamp the fold was given (who acted + under whose permission)
//   staged    the ordered off-bus input this step consumed (5.4)
//   actions   what was ASKED — differs from `results` whenever the gate refused
//   results   POST-GATE — exactly what was FOLDED
//   commands  the signed record, with the minted ids that cannot be re-derived
//   context   promptVersion + the rendered digest the model saw (G15/14.7)
//
// THE SCHEMA ENVELOPE (14.7). `schemaVersion` has NO DEFAULT: the one site that
// mints a record has to stamp it, and a historical record — [StepRecordV1] — is
// a DIFFERENT TYPE that `Replay.refold` cannot be handed at all. Upcasting is
// the only way an old log reaches the fold, which is exactly what 14.7 asks
// for: never touch history (14.1); transform on the way in.
//
// Kotlin gets the PAYLOAD half of that refusal for free, and it is worth naming
// because the other port does not: [adr.contract.TriageV1Result] does not extend
// `ToolResult`, so a v1 payload cannot be placed into `results` by hand no matter
// what the envelope says. The TypeScript port is structural and had to buy the
// same guarantee with a conflicting discriminant.

package adr.spine.pure

import adr.contract.Command
import adr.contract.ToolResult

/**
 * Which shape a committed record is in (14.7).
 *
 * A value class, like every other id in this port, so an Int cannot wander into
 * the field by accident. It is deliberately NOT the reducer version a snapshot is
 * tagged with (`reducerVersion`, declared at `app/Wire.kt`), and NOT the spine
 * version marker that says which template copy this tier is ([SPINE_VERSION],
 * declared at `spine/pure/Version.kt`): three independent questions, three
 * independent numbers, and the ratified record refuses to merge any two of them.
 */
@JvmInline
value class SchemaVersion(val value: Int)

/** The shape this port WRITES today. */
val CURRENT_SCHEMA = SchemaVersion(2)

/**
 * GENESIS is 1, and there is no v0: the reference persists no log, so the first
 * shape it ever wrote is the first shape anything can read. v1 survives as
 * [StepRecordV1] plus a fixture, so the upcast path is exercised rather than
 * described — and it is STAMPED, because a version an old record does not carry is
 * a version nothing could ever have dispatched on.
 */
val GENESIS_SCHEMA = SchemaVersion(1)

data class StepRecord(
    val schemaVersion: SchemaVersion,
    val now: Timestamp,
    val sig: Signature,
    val staged: List<StagedInput>,
    val actions: List<Action>,
    val results: List<ToolResult>,
    val commands: List<Command>,
    val context: ContextFixture,
)

/**
 * THE v1 ENVELOPE — what this port wrote while the version was 1.
 *
 * Generic in its payload, and that is G11 rather than taste: the payload shapes a
 * v1 log holds are a BLOCK's, and the spine may not name a block (gate check C15).
 * So the spine declares the envelope and the app supplies what its old payloads
 * were. `commands` is NOT generic, because Commands do not evolve here — a Command
 * is the signed record of what a principal authorized, and 14.1 forbids rewriting
 * it; the TS port makes the same call for the same reason.
 *
 * It is a separate type rather than a nullable field on [StepRecord], which is the
 * whole point: the two are different types, and only one of them re-folds. Its own
 * `schemaVersion` carries NO DEFAULT either — a fixture that could omit the version
 * would stop standing in for a record that genuinely carried one.
 */
data class StepRecordV1<R>(
    val schemaVersion: SchemaVersion,
    val now: Timestamp,
    val sig: Signature,
    val staged: List<StagedInput>,
    val actions: List<Action>,
    val results: List<R>,
    val commands: List<Command>,
    val context: ContextFixture,
)

/**
 * UPCAST ON READ (14.7): re-stamp the envelope, and lift every payload with the
 * app's own upcaster.
 *
 * A CONSTRUCTED type holding the payload lift, exactly as [adr.spine.replay.Replay]
 * holds the fold: the lift is fixed for one app and varies per call for nothing, so
 * the split rule puts it in the constructor and it drops out of the signature. It
 * touches no history — every call produces a NEW record, which is the difference
 * between upcasting and the rewrite 14.1 forbids.
 */
class SchemaUpcast<R>(private val payload: UpcastResult<R>) {

    fun v1(record: StepRecordV1<R>): StepRecord = StepRecord(
        schemaVersion = CURRENT_SCHEMA,
        now = record.now,
        sig = record.sig,
        staged = record.staged,
        actions = record.actions,
        results = record.results.map { payload(it) },
        commands = record.commands,
        context = record.context,
    )
}

File: agent-driven-architecture-main/examples/typescript/src/spine/pure/step-record.ts
// ── spine/pure/step-record — THE unit of commit and THE unit of replay (G9) ─
// The shipped reference committed a PAIR — append(signedCommands, results) —
// with no clock on it. Measured consequence: a live boundary folded at
// now = 1001 and its re-fold produced at = 0. Every timestamp was lost, and in
// any domain where `now` lands in State, the state was lost outright.
//
// So the commit is the STEP, not the pair. Every field earns its place:
//   schemaVersion  which shape the rest of the record is in (14.7) — the
//             envelope, so a log written across deployments is self-describing
//   now       without it a re-fold cannot reproduce what a live boundary wrote
//   sig       the stamp the fold was given
//   staged    the ordered off-bus input fixture this step consumed (5.4)
//   actions   what was ASKED — the audit half G1 named
//   results   POST-GATE — exactly what was FOLDED (so a refusal re-folds
//             without calling the authorization seam again: G9)
//   commands  the signed record, with ids that cannot be re-derived
//   context   promptVersion + the rendered digest the model saw (G15)
//
// THE SCHEMA ENVELOPE (14.7). `schemaVersion` is REQUIRED and typed as the
// LITERAL current version, so the compiler is the enforcement layer twice over:
// the boundary cannot commit a record without stamping one, and a HISTORICAL
// record — `StepRecordV1` below — is not a `StepRecord` at all and cannot be
// handed to `refold`. Upcasting is the only way in, which is exactly what 14.7
// asks for: never touch history (14.1); transform an old record into the
// current shape on the way into the fold.
//
// BOTH HALVES OF THE OLD RECORD ARE REFUSED, and the second half is the one a
// structural language makes easy to get wrong. The ENVELOPE half is this file's
// literal type. The PAYLOAD half belongs to the block: a historical payload has
// to CONFLICT with `ToolResultBase` on a member, or TypeScript's structural
// assignability would let it be spread by hand into a current-shape record and
// re-folded — a v2 envelope over v1 payloads, which is precisely the state the
// envelope exists to make unrepresentable. The triage block spells that
// conflict as `outcome: "ok-v1"`; see its contract for why a marker property
// would not have worked.

import type { Signature } from "./actor";
import type { SealedCommand } from "./command";
import type { ContextFixture } from "./context";
import type { RawInput, Timestamp, ToolName } from "./ids";
import type { StagedInput } from "./staged";
import type { SealedResult, ToolResultBase } from "./tool-result";
import { seal } from "./tool-result";

/** An OPEN boundary input (the 6.10 carve-out): a name and an undecoded blob.
 *  Declared here because it is a field of the committed record; re-exported
 *  from `spine/boundary/action`, where it is resolved. */
export interface Action {
  readonly tool: ToolName;
  readonly input: RawInput;
}

/** The shape this port WRITES today.
 *
 *  It is deliberately NOT the reducer version a snapshot is tagged with
 *  (`reducerVersion`, declared at `app/wire`), and NOT the spine version marker
 *  that says which template copy this tier is (`SPINE_VERSION`, declared at
 *  `spine/pure/version`). Three independent questions, three independent
 *  numbers, and the ratified record refuses to merge any two of them. */
export const SCHEMA_VERSION = 2;

/** GENESIS is 1, and there is no v0: the reference persists no log, so the
 *  first shape it ever wrote is the first shape anything can read. v1 survives
 *  as a type and a fixture so the upcast path is exercised rather than
 *  described — and it is STAMPED, because a version an old record does not
 *  carry is a version nothing could ever have dispatched on. */
export const GENESIS_SCHEMA_VERSION = 1;

export interface StepRecord {
  /** The envelope (14.7). Typed as the literal, so a record in any other shape
   *  — including a correctly-shaped record still stamped GENESIS — is not a
   *  `StepRecord`, and the compiler rather than a runtime branch refuses it at
   *  the door of the fold. */
  readonly schemaVersion: typeof SCHEMA_VERSION;
  readonly now: Timestamp;
  readonly sig: Signature;
  readonly staged: readonly StagedInput[];
  readonly actions: readonly Action[];
  /** SEALED, and that is the second half of C7 (spine/pure/tool-result). A
   *  result reaches this list from the boundary's own map or from 14.7's
   *  upcast below and from nowhere else — `{ ...received }` in a fold arm
   *  produces a value this field will not take. */
  readonly results: readonly SealedResult[];
  readonly commands: readonly SealedCommand[];
  readonly context: ContextFixture;
}

/** THE v1 ENVELOPE — what this port wrote while the version was 1.
 *
 *  Generic in its payload, and that is G11 rather than taste: the payload
 *  shapes a v1 log holds are a BLOCK's, and the spine may not name a block
 *  (check C15). So the spine declares the envelope and the app supplies what
 *  its old payloads were. `commands` is NOT generic, because Commands do not
 *  evolve here — a Command is the signed record of what a principal
 *  authorized, and 14.1 forbids rewriting it. (In this port a block's contract
 *  may not even NAME a Signature — check C4 — so a historical Command shape
 *  has nowhere it could be declared.)
 *
 *  It is a separate interface rather than a `Partial<StepRecord>`: the point of
 *  the envelope is that the two are DIFFERENT TYPES, one of which is refused. */
export interface StepRecordV1<R> {
  readonly schemaVersion: typeof GENESIS_SCHEMA_VERSION;
  readonly now: Timestamp;
  readonly sig: Signature;
  readonly staged: readonly StagedInput[];
  readonly actions: readonly Action[];
  readonly results: readonly R[];
  readonly commands: readonly SealedCommand[];
  readonly context: ContextFixture;
}

/** UPCAST ON READ (14.7): re-stamp the envelope, and lift every payload with
 *  the app's own upcaster. History is not touched — a NEW value is produced on
 *  the way into the fold, which is the whole difference between upcasting and
 *  the rewrite 14.1 forbids. */
export function upcastV1<R>(
  record: StepRecordV1<R>,
  payload: (result: R) => ToolResultBase,
): StepRecord {
  return {
    ...record,
    schemaVersion: SCHEMA_VERSION,
    // THE SECOND MINT SITE, and the only one outside `spine/boundary`. A
    // block's upcaster produces the PAYLOAD (check C7 puts it in the block's
    // tools.ts); the seal is the spine's, applied here, because this is the one
    // path by which a record written before the current shape existed reaches
    // the fold. Sealing at the boundary instead would be unreachable: an old
    // record never passes through the live name→result map.
    results: record.results.map((result) => seal(payload(result))),
  };
}
